### ISSUE TITLE
Drop avatars from attachments

### ISSUE THREAD

**Issue by emanuele45** [elkarte/Elkarte#2486]
Something I have in mind, but never tracked, I would really see at some point the option to upload avatars as attachments to go away.
Attachments should be just that, not mixture of stuff.
Avatars should just be uploaded to a directory and directly accessible from the outside world.

Of course that would require attention to possible tricky files (re-encoding should be a must... even though that would make animated gif a little hard to deal with unless we can rely on Image Magick), but it would clean up a lot of mess. IMO.

**Comment by Frenzie** [comment]
Not just animated GIFs. You'd probably increase the file size of just about any file I'm likely to attach. To quote [a bit of my attachment philosophy](http://blogs.opera.com/desktop/2013/10/welcome-to-the-new-home-of-desktop-blog/#comment-1106162856) (but it's equally applicable to avatars).

> A screenshot of your comment is 16kB in PNG. Subsequently running pngquant gives a 6.2kB result with no discernible loss in quality. A libjpeg optimized JPEG looks alright at 14-18kB, but it has some relatively minor artifacts. You can easily reach about 6kB with JPEG as well, but the artifacts then become really distracting to the point where it looks more like one of those illegible CAPTCHAs.
>
> These results indicate that *the system should check how filesize compares between e.g. carefully crafted PNGs and its own low-quality JPEGs, otherwise it shoots itself in the foot* (emphasis added). For that matter, it really shouldn't add another layer of noise to a properly optimized JPEG either. In the past I've uploaded a JPEG to Facebook, which consequently made the file bulkier and uglier.
>
> Above all, however, this blog is not Facebook. Like I indicated before, screenshots are doubtless the primary use case, not photos. I'd say simply reject all images over 200kB, or possibly even as little as 50kB. Give some pointers about how to optimize files, but don't let the system muck about with it. I know, tough love, but I see no added value in shrunk low-quality JPEG images.

**Comment by emanuele45** [comment]
Well, it's not that you have an infinite amount of avatars, and usually are just tiny crappy images anyway, not 5Mpx photo, just tiny 150x150px simple images of about... 20kb? Multiplied, in the worst case, by the number of members of your board. And let's not forget that if the avatar is bigger than the defined size, it is resized (with GD or IM), so it doesn't really change that much, unless you resize all the avatars of your members... are you?

Have avatars in the attachments system is a P.I.T.A. because it forces us to query the database any time we need to know the avatar of a user, having the avatars directly accessible from the outside with a defined naming scheme (e.g. {id_member}_avatar_b.png, {id_member}_avatar_m.png, {id_member}_avatar_s.png for big, medium, small) would give us the option to even forget about the img tag and use instead CSS for avatars as well (no idea if it is worth, just an option).

**Comment by Frenzie** [comment]
Oh, there really shouldn't be any special DB queries on a per avatar basis. That's just stupid. ;) The only thing that makes sense is something like a `has_avatar` column for a little check where avatars are displayed. 

On your 150x150px, I'd say that with "HiDPI" monitors being standard now[1] that should really be x2 with e.g. `srcset` and `sizes` (and some kind of JS polyfill for browsers that don't support this stuff natively). Generating the "regular" size image is something the server should probably do. My concern with image size is page loading time and (mobile) data waste, not server storage space.

Taking my own avatar with a purposefully reduced color palette (for artsy reasons) on ElkArte.net as an example.

![avatar-fs8](https://cloud.githubusercontent.com/assets/202757/14317384/4ac39de0-fc08-11e5-93cd-2d5f7ee204a1.png)

The regular `optipng -o7` PNG is 15.8 kiB while the pngquant version is 8.2 kiB. Resaving it with ImageMagick results in 15.8 kiB again. So even if everyone had my minuscule avatar, that might easily approach over 200 kiB extra per forum page, because who goes with the extremely low default of 15 posts per page anyway? =P

```
$ du -h avatar.png 
16K	avatar.png
$ pngquant avatar.png
$ du -h avatar-fs8.png 
8.0K	avatar-fs8.png
$ convert avatar-fs8.png avatar-resaved.png
$ du -h avatar-resaved.png 
16K	avatar-resaved.png
```

And since JPEG results in quality loss, imo as a rule of thumb only PNG is acceptable and nothing should be done if the resulting image after processing is of a bigger size than the original. Unfortunately, as far as I'm aware PHP only typically comes with ImageMagick modules and nothing like OptiPNG. pngquant is also unacceptable because that is (or easily can be) lossy, and it's not available anyway.

tl;dr Some people optimize their images and undoing their work Facebook style is like a slap in the face. Even more so because ElkArte, like its relative Wedge, is "like Facebook, but not for dummies" ([source](https://github.com/Wedge/wedge)).

[1] I might be getting ahead of myself there, but I've had mine for over a year and it's fantastic.

**Comment by Spuds** [comment]
Yeah I really wanted to eliminate the db query stuff in 1.0 but there were bigger problems to fix.

I'd like to see Avatars with a thumbnail (120x120?) and normal (250x250?) sizes, just like attachments.  The thumb is used (and css smaller ones of that thumb) as needed in most places, board index, etc
In some areas, like profile, the thumb links to the larger image, like Github does as well as some other forum packages.

The file-size is an interesting item ... could check to see if the created is > what was uploaded and act accordingly but I can see some issues with that as well.

**Comment by Spuds** [comment]
This was addressed in 2.0
